 # **Project Name** - **Booking.com - Hotel Booking Analysis**

##### **Project Type**    - EDA
##### **Contribution**    - Individual

# **Problem Statement**

##### Booking.com manages a wide range of hotel bookings across various locations, customer profiles, and channels. The dataset includes key details like booking lead times, room types, customer preferences, and reservation statuses. The challenge is to analyze how factors like booking windows, guest demographics, and reservation types affect hotel performance and customer experience. By leveraging these insights, Booking.com aims to **optimize booking efficiency, predict guest needs, and improve customer satisfaction, ensuring a competitive edge in the dynamic hospitality industry**

# **General Guidelines :**

#***Let's Begin !***

## ***1. Know Your Data***

### Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### Dataset Loading

In [2]:
df = pd.read_csv('/content/drive/MyDrive/Local Disk (D:)/Data Analyst/Project_Root/Data/Raw/Hotel Bookings.csv')

### Dataset First View

### Dataset Rows & Columns count

In [5]:
df.shape

(119390, 32)

### Dataset Information

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [7]:
null_values = df.isnull().sum()
columns_with_nulls = null_values[null_values > 0]
print(columns_with_nulls)

children         4
country        488
agent        16340
company     112593
dtype: int64


### What did you know about your dataset?

* Booking offers two types of hostels across 177 countries. The dataset contains 119390 rows and 32 columns. Four columns—children, country, agent, and company—have missing/NaN values, with the company column having the highest number of NaN values.*

In [9]:
%cd /content/drive/MyDrive/Local Disk (D:)/Data Analyst/Project_Root

/content/drive/MyDrive/Local Disk (D:)/Data Analyst/Project_Root


In [15]:
!git add .
!git commit -m 'Know your data - completed'


[master 2886403] Know your data - completed
 1 file changed, 1 insertion(+), 1 deletion(-)
 rewrite Notebooks/Booking.com - Hotel Booking Analysis.ipynb (93%)


Enumerating objects: 22, done.
Counting objects: 100% (22/22), done.
Delta compression using up to 2 threads
Compressing objects: 100% (17/17), done.
Writing objects: 100% (19/19), 1.23 MiB | 1.47 MiB/s, done.
Total 19 (delta 3), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (3/3), done.
To https://github.com/kaushdotin/Project_Root.git
   e43bcd5..2886403  master -> master


## ***2. Understanding Your Variables***

In [212]:
df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

#### Variables Description

* **hotel** : hotel type (City Hotel, Resort Hotel)
* **is_canceled :** Booking canceled(1) and if not canceled=(0)
* **lead_time :** Elapsed days between entering date of booking and arrival date
* **arrival_date_year :** Year of arrival
* **arrival_date_month :** Month of arrival
* **arrival_date_week_number :** Week number of arrival date
* **arrival_date_day_of_month :** Date of arrival
* **stays_in_weekend_nights :** Number of weekend night (Saturday or Sunday) the guest stayed or booked in the hotel
* **stays_in_week_nights :** Number of weekday night (Monday to Friday) the guest stayed or booked in the hotel
* **adults :** Number of adults
* **children :** Number of children
* **babies :** Number of babies
* **meal :** Kind of mean
* **country :** Country code
* **market_segment :** Which segment customer belong to
* **distribution_channel :** How customer assessed stay - corporate booking/Direct/TA.TO
* **is_repeated_guest :** If guest coming first time(1) else (0)
* **previous_cancellations :** Was there cancelation before
* **previous_bookings_not_canceled :** Previous bookings was not canceled
* **reserved_room_type :** Types of room reserved
* **assigned_room_type :** Type of room assined to guest
* **booking_changes :** If there was any change made in booking
* **deposit_type :** Type of deposit - ['No Deposit', 'Refundable', 'Non Refund']
* **agent :** Booking through agent
* **company :** Company Name
* **days_in_waiting_list :** Number of days in waitlist
* **customer_type :** Type of customer
* **adr :** Average daily rate
* **required_car_parking_spaces :** If car parking is required
* **total_of_special_requests :** Number of special requirement
* **reservation_status :** Reservation status
* **reservation_status_date :** Reservation status date

In [4]:
df.describe(include='all')

hotel    is_canceled      lead_time  arrival_date_year  \
count       119390  119390.000000  119390.000000      119390.000000   
unique           2            NaN            NaN                NaN   
top     City Hotel            NaN            NaN                NaN   
freq         79330            NaN            NaN                NaN   
mean           NaN       0.370416     104.011416        2016.156554   
std            NaN       0.482918     106.863097           0.707476   
min            NaN       0.000000       0.000000        2015.000000   
25%            NaN       0.000000      18.000000        2016.000000   
50%            NaN       0.000000      69.000000        2016.000000   
75%            NaN       1.000000     160.000000        2017.000000   
max            NaN       1.000000     737.000000        2017.000000   

       arrival_date_month  arrival_date_week_number  \
count              119390             119390.000000   
unique                 12                       NaN   
top                August                       NaN   
freq                13877                       NaN   
mean                  NaN                 27.165173   
std                   NaN                 13.605138   
min                   NaN                  1.000000   
25%                   NaN                 16.000000   
50%                   NaN                 28.000000   
75%                   NaN                 38.000000   
max                   NaN                 53.000000   

        arrival_date_day_of_month  stays_in_weekend_nights  \
count               119390.000000            119390.000000   
unique                        NaN                      NaN   
top                           NaN                      NaN   
freq                          NaN                      NaN   
mean                    15.798241                 0.927599   
std                      8.780829                 0.998613   
min                      1.000000                 0.000000   
25%                      8.000000                 0.000000   
50%                     16.000000                 1.000000   
75%                     23.000000                 2.000000   
max                     31.000000                19.000000   

        stays_in_week_nights         adults  ...  deposit_type          agent  \
count          119390.000000  119390.000000  ...        119390  103050.000000   
unique                   NaN            NaN  ...             3            NaN   
top                      NaN            NaN  ...    No Deposit            NaN   
freq                     NaN            NaN  ...        104641            NaN   
mean                2.500302       1.856403  ...           NaN      86.693382   
std                 1.908286       0.579261  ...           NaN     110.774548   
min                 0.000000       0.000000  ...           NaN       1.000000   
25%                 1.000000       2.000000  ...           NaN       9.000000   
50%                 2.000000       2.000000  ...           NaN      14.000000   
75%                 3.000000       2.000000  ...           NaN     229.000000   
max                50.000000      55.000000  ...           NaN     535.000000   

            company days_in_waiting_list customer_type            adr  \
count   6797.000000        119390.000000        119390  119390.000000   
unique          NaN                  NaN             4            NaN   
top             NaN                  NaN     Transient            NaN   
freq            NaN                  NaN         89613            NaN   
mean     189.266735             2.321149           NaN     101.831122   
std      131.655015            17.594721           NaN      50.535790   
min        6.000000             0.000000           NaN      -6.380000   
25%       62.000000             0.000000           NaN      69.290000   
50%      179.000000             0.000000           NaN      94.575000   
75%      270.000000             0.000000           

#### Unique values in each variables

In [101]:
for i in df.columns.tolist():
  print(f"Unique value for {i} : { df[i].nunique()}.")

Unique value for hotel : 2.
Unique value for is_canceled : 2.
Unique value for lead_time : 479.
Unique value for arrival_date_year : 3.
Unique value for arrival_date_month : 12.
Unique value for arrival_date_week_number : 53.
Unique value for arrival_date_day_of_month : 31.
Unique value for stays_in_weekend_nights : 17.
Unique value for stays_in_week_nights : 35.
Unique value for adults : 14.
Unique value for children : 5.
Unique value for babies : 5.
Unique value for meal : 5.
Unique value for country : 177.
Unique value for market_segment : 8.
Unique value for distribution_channel : 5.
Unique value for is_repeated_guest : 2.
Unique value for previous_cancellations : 15.
Unique value for previous_bookings_not_canceled : 73.
Unique value for reserved_room_type : 10.
Unique value for assigned_room_type : 12.
Unique value for booking_changes : 21.
Unique value for deposit_type : 3.
Unique value for agent : 333.
Unique value for company : 352.
Unique value for days_in_waiting_list : 128.


In [214]:
!git add .
!git commit -m "Understanding of variable"

On branch master
nothing to commit, working tree clean


## ***3. Data Wrangling***

In [6]:
# Deleting company column as it has heighest NaN values
df.drop(columns="company", inplace=True)

In [7]:
# Dropping NaN values
df.dropna(inplace=True)

In [8]:
df.shape

(102894, 31)

In [5]:
null_values = df.columns.isnull().any()
null_values

False

When > Who > Why > How to improve

In [9]:
# Store canceled bookings hotel wise data
canceled_bookings_hotel_type = canceled_bookings.groupby('hotel').size()

# Total bookings
total_booking_hotel_type = df.groupby('hotel').size()

# Create DataFrame boookings and canceled booking
booking_hotel_type_df = total_booking_hotel_type.reset_index('hotel')
booking_hotel_type_df.columns=['hotel', 'booked']


booking_hotel_type_df['canceled'] = canceled_bookings_hotel_type.values
booking_hotel_type_df['percentage'] = booking_hotel_type_df['canceled'] / booking_hotel_type_df['booked'] * 100

booking_hotel_type_df

hotel  booked  canceled  percentage
0    City Hotel   79330     33102   41.726963
1  Resort Hotel   40060     11122   27.763355

In [352]:
# Bookings canceled by country :

bookings_by_country = df.groupby('country').size().sort_values(ascending=False)
canceled_bookings_by_country = canceled_bookings.groupby('country').size().sort_values(ascending=False)

bookings_by_country_df = bookings_by_country.reset_index('country')
bookings_by_country_df.columns = ['country', 'total_bookings']

bookings_by_country_df['canceled'] = bookings_by_country_df['country'].map(canceled_bookings_by_country)

bookings_by_country_df['percetange'] = bookings_by_country_df['canceled'] / bookings_by_country_df['total_bookings'] * 100

top_10_country_with_heighest_bookings = bookings_by_country_df.sort_values(['canceled','percetange'], ascending=False).head(10)

# Here is the to 10 country which has maximum canceled bookings :
top_10_country_with_heighest_bookings

country  total_bookings  canceled  percetange
0     PRT           37591   23725.0   63.113511
1     GBR           10889    2353.0   21.608963
3     ESP            7848    2168.0   27.624873
2     FRA            9810    1897.0   19.337411
5     ITA            3555    1331.0   37.440225
4     DEU            6998    1212.0   17.319234
6     IRL            3198     831.0   25.984991
8     BRA            2086     827.0   39.645254
9     USA            1970     498.0   25.279188
7     BEL            2156     469.0   21.753247

In [353]:
# Bookings canceled by guest type :
canceled_by_guest = canceled_bookings.groupby('is_repeated_guest').size()

bookings_canceled_by_repeated_guest_percentage = canceled_by_guest[1] / len(canceled_bookings) * 100
bookings_canceled_by_new_guest_percentage = canceled_by_guest[0] / len(canceled_bookings) * 100

print(f"Booking canceled by new guest : {bookings_canceled_by_new_guest_percentage}%")
print(f"Booking canceled by repeated guest : {bookings_canceled_by_repeated_guest_percentage}%")

Booking canceled by new guest : 99.03637857623067%
Booking canceled by repeated guest : 0.9636214237693284%


In [290]:
# Booking canceled by market segmnet :
canceled_by_market_segment = canceled_bookings.groupby('market_segment').size().sort_values(ascending=False)
canceled_by_market_segment

market_segment
Online TA        20693
Groups           10059
Offline TA/TO     8041
Direct            1179
Corporate          171
Complementary       12
Aviation             6
dtype: int64

In [10]:
# Define a reusable function to count canceled bookings based on stay type
def count_canceled_bookings_by_stay_type(df, stay_column):
    filtered_df = df[df[stay_column] > 0]
    return filtered_df[stay_column].count()

# Count canceled bookings by weekend stays :
bookings_canceled_by_weekend_stays = count_canceled_bookings_by_stay_type(canceled_bookings, 'stays_in_weekend_nights')

# Count canceled bookings by week night stays :
bookings_canceled_by_week_night_stays = count_canceled_bookings_by_stay_type(canceled_bookings, 'stays_in_week_nights')

# Print the results :

print(f'Bookings canceled by weekend stays: {bookings_canceled_by_weekend_stays}')
print(f'Bookings canceled by week night stays: {bookings_canceled_by_week_night_stays}')

Bookings canceled by weekend stays: 24723
Bookings canceled by week night stays: 42311


In [349]:
canceled_bookings_reserved_room_type = canceled_bookings.groupby('reserved_room_type').size()
canceled_bookings_assinged_room_type = canceled_bookings.groupby('assigned_room_type').size()

canceled_bookings_assinged_room_typex = canceled_bookings.groupby('required_car_parking_spaces').size()
canceled_bookings_assinged_room_typex

required_car_parking_spaces
0    40161
dtype: int64

In [295]:
df

hotel  is_canceled  lead_time  arrival_date_year  \
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
5       Resort Hotel            0         14               2015   
7       Resort Hotel            0          9               2015   
8       Resort Hotel            1         85               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
3                    July                        27   
4                    July                        27   
5                    July                        27   
7                    July                        27   
8                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
3                               1                        0   
4                               1                        0   
5                               1                        0   
7                               1                        0   
8                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  booking_changes  deposit_type  \
3                          1       1  ...                0    No Deposit   
4                          2       2  ...                0    No Deposit   
5                          2       2  ...                0    No Deposit   
7                          2       2  ...                0    No Deposit   
8                          3       2  ...                0    No Deposit   
...                      ...     ...  ...              ...           ...   
119385                     5       2  ...                0    No Deposit   
119386                     5       3  ...                0    No Deposit   
119387                     5       2  ...                0    No Deposit   
119388                     5       2  ...                0    No Deposit   
119389                     7       2  ...                0    No Deposit   

        agent days_in_waiting_list customer_type     adr  \
3       304.0                    0     Transient   75.00   
4       240.0                    0     Transient   98.00   
5       240.0                    0     Transient   98.00   
7       303.0                    0     Transient  103.00   
8       240.0                    0     Transient   82.00   
...       ...                  ...           ...     ...   
119385  394.0                    0     Transient   96.14   
119386    9.0                    0     Transient  225.43   
119387    9.0                    0     Transient  157.71   
119388   89.0                    0     Transient  104.40   
119389    9.0                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
3                                 0                          0   
4                                 0   

In [244]:
# Finding IQR

percentile25 = bookings_by_country_df['canceled'].quantile(0.25)
percentile75 = bookings_by_country_df['canceled'].quantile(0.75)

In [245]:
# Calculate IQR

iqr = percentile75 - percentile25

In [257]:
upper_limit = percentile75 + 1.5 * iqr
lower_limit = percentile25 - 1.5 * iqr

upper_limit
# lower_limit

95.5

## ***4. Data Vizualization, Storytelling & Experimenting with charts : Understand the relationships between variables***

## ***5. Solution to Business Objective***

## ***Conclusion***

In [268]:
df

hotel  is_canceled  lead_time  arrival_date_year  \
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
5       Resort Hotel            0         14               2015   
7       Resort Hotel            0          9               2015   
8       Resort Hotel            1         85               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
3                    July                        27   
4                    July                        27   
5                    July                        27   
7                    July                        27   
8                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
3                               1                        0   
4                               1                        0   
5                               1                        0   
7                               1                        0   
8                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  booking_changes  deposit_type  \
3                          1       1  ...                0    No Deposit   
4                          2       2  ...                0    No Deposit   
5                          2       2  ...                0    No Deposit   
7                          2       2  ...                0    No Deposit   
8                          3       2  ...                0    No Deposit   
...                      ...     ...  ...              ...           ...   
119385                     5       2  ...                0    No Deposit   
119386                     5       3  ...                0    No Deposit   
119387                     5       2  ...                0    No Deposit   
119388                     5       2  ...                0    No Deposit   
119389                     7       2  ...                0    No Deposit   

        agent days_in_waiting_list customer_type     adr  \
3       304.0                    0     Transient   75.00   
4       240.0                    0     Transient   98.00   
5       240.0                    0     Transient   98.00   
7       303.0                    0     Transient  103.00   
8       240.0                    0     Transient   82.00   
...       ...                  ...           ...     ...   
119385  394.0                    0     Transient   96.14   
119386    9.0                    0     Transient  225.43   
119387    9.0                    0     Transient  157.71   
119388   89.0                    0     Transient  104.40   
119389    9.0                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
3                                 0                          0   
4                                 0   